### DISCUSSION TITLE
useApiKey for the entire collection

### DISCUSSION THREAD

**Discussion by gotleg** [payloadcms/payload#618] [Q&A]
Hello,

I'm testing payloadcms for a few hours and so far, everything is pretty straight forward. Nice job :)

I read the doc about enabling api-key auth on a collection, but once it's done, I have to generate an api-key for each document !?!
Is there a way to generate an api-key for one user but on an entire collection and even better for multiple collections ?

I want to be able to list all items in a collection (ex: /api/menu)

Can't find a way to do so in the docs. Do I miss something ?

Thanks for your help

**Comment by jmikrut** [comment]
Hey @gotleg — thank you for the compliments!

I am not sure I understand your questions. You _**do**_ have to generate an API key for each _**user**_ that requires one, but I am not sure why you'd need each user to instantly have an API key. Users can _**log in**_ via JWT without generating an API key, which is the normal way for users to authenticate, and API keys are more of a way for your third-party apps and integrations to be able to auth with Payload directly without having to log in before making requests. Typically you only have a small handful of API keys (maybe one for each service that is integrating with Payload). 

But, if you wanted to auto-generate an API key for each user in a collection (or multiple collections), you could by simply writing a Node script that uses Payload's Local API to update all users with `enableAPIKey: true` and then a generated API key that you set.

Let me know if I am misunderstanding here. Thanks again!

**Reply by gotleg** [reply]
Hi @jmikrut,

I think I understood my mistake and it's more logical.
I understood from the docs that enableAPIKey had to be placed on each collection that needs to be accessed via a client app to enable API-Key authentication on this particular collection. 
Thanks to your answer, I realized that enableAPIKey has to be placed on Users collection as an optionnal auth method for each user.

This way it's exactly what I was looking for :)

I just have to create a specific user for my client APP and use its generated API-Key as Auth method in my Http requests :D

I'm used to place http header for Authorization like this : 
Authorization: Bearer xxxxxxxxxxxxxxxxxxxxxxx

so according to the doc it should be for payloadcms : 
Authorization: API-Key xxxxxxxxxxxxxxxxxxxxxxx
https://payloadcms.com/docs/authentication/config#api-keys

Is that correct ?

Thanks for your help.

**Reply by JarrodMFlesch** [reply]
@Gotleg it would actually be

```
Authorization: User API-Key xxxxxxxxxxxxxxxxxxxxxxx
```

Assuming your auth collections **Singular Label** is **User**


You could also have other api keys on other collections,

```
Authorization: Admin API-Key xxxxxxxxxxxxxxxxxxxxxxx
```

Assuming this auth collections **Singular Label** is **Admin**

**Reply by gotleg** [reply]
Perfect thank you

**Reply by akanshSirohi** [reply]
@gotleg can you please tell how you generated api-key for each collection and integrated it..? 
As there is no such example or steps in the documentation.

**Reply by gotleg** [reply]
@akanshSirohi So the process is pretty simple.
You have to activate api-key authentication on your user collection to enable api-key auth instead of JWT auth.
`
const Users: CollectionConfig = {
  slug: 'users',
  auth: {
    useAPIKey: true
  },
...
`

Then you can activate api-key auth on a user and generate the api-key : 
![image](https://user-images.githubusercontent.com/2339335/173043809-615f215b-7ccd-4c86-94cd-c8df983cbb2e.png)

**Reply by akanshSirohi** [reply]
@gotleg Thanks for saving my day! 😄

**Comment by drewbo** [comment]
This pattern still seems a bit off to me but please feel free to correct my understanding if I'm way off-base (I'm testing using the website template):
- setting `useAPIKey: true` at the collection level adds a new component to the admin panel for each collection element, where I can add an API token for a given user
- that token is visible to all users with admin access (even if it doesn't match their email)
- the token doesn't show up in other elements of the same collection (i.e. it seems element specific)
- the token isn't visible in the User collection as being associated with a single user

A more common pattern I've seen with API tokens (tailored to this situation):
- assigned to a single-user (still potentially an "integration user" as suggested by the docs)
- the token is assigned permissions (e.g. read/write access to `posts`, read only to `users`)
- the token wouldn't necessarily be visible at the collection level or individual element level, especially not for users whose email it doesn't match (or gate to a certain level of access) but it should be seen for a given user.

**Reply by drewbo** [reply]
Hi! I'm coming back to correct myself. It does work somewhat as expected!

My main point of confusion was that the [example in the documentation](https://payloadcms.com/docs/authentication/api-keys) doesn't show the `auth` property being set on the `User` collection, which seems like the standard way this would be used. The example worked out by the original poster is a bit more clear.

One additional question: is it possible to hide the API key for other users (or otherwise control access to the key) given that it isn't specified as a field?
